### PR TITLE
Implement the delegation mechanism

### PR DESCRIPTION
**Pull request by hihiben**
A `delegatee` is allowed to execute on a user's behalf.

- [x] Implement allow
  - [x] allow
  - [x] allowBySig
- [x] Implement disallow
- [x] Tests

### CODE REVIEW

File: src/Router.sol [hunk 1]
```diff
@@ -29,6 +31,9 @@ contract Router is IRouter, EIP712, FeeGenerator {
     /// @notice Mapping for recording exclusive agent contract to each user
     mapping(address user => IAgent agent) public agents;
 
+    /// @notice Mapping for user of each delegatee
+    mapping(address delegatee => address user) public delegators;
+
```
Comment (zodahu) [inline src/Router.sol:36]: Does this mapping indicate that one delegatee can only serve one user? It seems that other users can point the delegatee to other users through allow.

Comment (hihiben) [reply src/Router.sol:36]: Yes I suppose that should be the case. In this scenario, the delegatee will be the one to call router.
- If `user` be the key -> `user` can only have one `delegatee`, `delegatee` can serve multiple `users` -> `delegatee` owns multiple authorities which might need addithional permission design  
- If `delegatee` be the key -> `delegatee` can only serve one `user`, `user` can have multiple delegatee -> the `delegatee` execution only relates to a single `user`

**Comment by ksin751119** [comment]
I have two questions:

1. In this design, does it align with the Web3Function service for one delegator to serve one user? Or would it be more beneficial to use one delegator to serve multiple users, considering that there might be limitations on the trigger amount for Web3Function?

2. Have you considered using operation signatures as a replacement for the allowing mechanism? With the signature representing the permitted operation for users, they could include the signature in Web3Function to indicate a specific operation, rather than allowing all operations for the delegator.

**Comment by hihiben** [comment]
> 1. In this design, does it align with the Web3Function service for one delegator to serve one user? Or would it be more beneficial to use one delegator to serve multiple users, considering that there might be limitations on the trigger amount for Web3Function?
For the integration of Automate and W3F, the delegated msg.sender will be different between every task creator, which is fine on this design. Also, if multiple users share the same delegatee, we will need an extra mechanism to differentiate the execution between different user. This requires not only an extra parameter representing the user, but also something like signature to make sure that the action is assigned by the correct user.
> 2. Have you considered using operation signatures as a replacement for the allowing mechanism? With the signature representing the permitted operation for users, they could include the signature in Web3Function to indicate a specific operation, rather than allowing all operations for the delegator.
Yes, but I choose to give the delegatee a higher trust to preserve the flexibility on possible action. If extra constraints are required due to safety concern, it can be define on the delegatee layer.

File: src/Router.sol [hunk 2]
```diff
@@ -232,6 +308,37 @@ contract Router is IRouter, EIP712, FeeGenerator {
...
+    }
+
+    /// @notice Set delegation information via signature
+    /// @param delegation The delegation information to be signed by a signer using EIP-712
+    /// @param signer The signer address
+    /// @param signature The signer's signature bytes
+    function allowBySig(IParam.Delegation calldata delegation, address signer, bytes calldata signature) external {
+        // Verify deadline, signer and signature
+        uint256 deadline = delegation.deadline;
+        if (block.timestamp > deadline) revert SignatureExpired(deadline);
```
Comment (zodahu) [inline src/Router.sol:325]: I'm thinking if we should check both deadline and nonce for allowBySig as it's critical unlike signer fee which aloows reused signature before deadline. What do you think?

Ref
- [Comet](https://github.com/compound-finance/comet/blob/main/contracts/CometExt.sol#L204) checks both nonce and deadline.
- [Permit2](https://github.com/Uniswap/permit2/blob/main/src/libraries/Allowance.sol#L26) packs nonce, expiry and amount to reduce gas.

File: src/Router.sol [hunk 3]
```diff
@@ -232,6 +308,37 @@ contract Router is IRouter, EIP712, FeeGenerator {
...
+
+    function _allow(address user, address delegatee, uint256 expiry) internal {
+        delegatees[user][delegatee] = expiry;
+        emit Delegated(user, delegatee, expiry);
+    }
+
+    /// @notice Disallow another address to execute on user's behalf
+    /// @param delegatee The address to be disallowed
+    function disallow(address delegatee) external {
+        delegatees[msg.sender][delegatee] = 0;
```
Comment (zodahu) [inline src/Router.sol:338]:
```suggestion
        delete delegatees[msg.sender][delegatee];
```

File: src/Router.sol (hunk 2 above)
Comment (hihiben) [reply src/Router.sol:325]: Indeed. Will do.

Comment (zodahu) [reply src/Router.sol:325]: @Jeff-CCH, Router's inheritance of EIP-712 prevents cross-chain replay.

File: src/Router.sol [hunk 4]
```diff
@@ -105,7 +107,7 @@ contract Router is IRouter, EIP712, FeeGenerator {
     /// @param user The delegator of the sender
     /// @return Valid or not
     function isValidDelegateeFor(address user) public view returns (bool) {
-        return block.timestamp <= delegatees[user][msg.sender];
+        return block.timestamp <= uint256(delegations[user][msg.sender].expiry);
```
Comment (Jeff-CCH) [inline src/Router.sol:110]: I'm wondering if it's a good idea to use `uint128` for expiry. Eventually, the block timestamp will be over uint128.max and we will have no valid delegatee.

Comment (hihiben) [reply src/Router.sol:110]: In practice it's not quite possible for timestamp to exceed `uint128`. [Permit2](https://github.com/Uniswap/permit2/blob/main/src/interfaces/IAllowanceTransfer.sol#L83C1-L83C1) library uses `uint48` to store expiration.

File: src/Router.sol [hunk 5]
```diff
@@ -232,6 +309,62 @@ contract Router is IRouter, EIP712, FeeGenerator {
...
     }
 
+    /// @notice Allow another address to execute on user's behalf
+    /// @param delegatee The address to be delegated
+    /// @param expiry The expiry of the delegation
+    function allow(address delegatee, uint128 expiry) external {
+        IParam.PackedDelegation storage delegation = delegations[msg.sender][delegatee];
+        delegation.updateExpiry(expiry);
+        emit Delegated(msg.sender, delegatee, expiry);
+    }
```
Comment (zodahu) [inline src/Router.sol:319]:
```suggestion
    function allow(address delegatee, uint128 expiry) public {
        delegations[msg.sender][delegatee].expiry = expiry;
        emit Delegated(msg.sender, delegatee, expiry);
    }
    
    function disallow(address delegatee) external {
        allow(delegatee, 0);
    }
```
Given the simplicity of the `updateExpiry` logic compared to Permit2's, prefer to have a consistent update method like `[delegatee].nonce = newNonce` as seen on line #364. Additionally, the disallow function can leverage the allow function.

File: test/Router.t.sol [hunk 6]
```diff
@@ -392,4 +398,178 @@ contract RouterTest is Test, LogicSignature {
...
+            '',
+            inputsEmpty,
+            IParam.WrapMode.NONE,
+            address(0), // approveTo
+            address(0) // callback
+        );
+        vm.prank(delegatee);
+        vm.expectRevert();
+        vm.warp(expiry + 1);
+        router.executeFor(user, logics, tokensReturnEmpty, SIGNER_REFERRAL);
```
Comment (zodahu) [inline test/Router.t.sol:503]:
```suggestion
        vm.prank(delegatee);
        vm.expectRevert();
        vm.warp(expiry + 1);
        router.executeFor(user, logicsEmpty, tokensReturnEmpty, SIGNER_REFERRAL);
```

File: test/Router.t.sol [hunk 7]
```diff
@@ -392,4 +398,178 @@ contract RouterTest is Test, LogicSignature {
...
+        logics[0] = IParam.Logic(
+            address(mockTo), // to
+            '',
+            inputsEmpty,
+            IParam.WrapMode.NONE,
+            address(0), // approveTo
+            address(0) // callback
+        );
+        vm.prank(delegatee);
+        router.executeFor(user, logics, tokensReturnEmpty, SIGNER_REFERRAL);
```
Comment (zodahu) [inline test/Router.t.sol:483]:
```suggestion
        vm.prank(delegatee);
        router.executeFor(user, logicsEmpty, tokensReturnEmpty, SIGNER_REFERRAL);
```

File: test/Router.t.sol [hunk 8]
```diff
@@ -392,4 +398,178 @@ contract RouterTest is Test, LogicSignature {
...
+    }
+
+    function testCannotAllowBySigWithIncorrectNonce() external {
+        // Ensure correct EIP-712 encodeData
+        uint256 deadline = block.timestamp + 3600;
+        uint128 expiry = uint128(deadline) + 3600;
+        uint128 nonce = 1;
+        IParam.DelegationDetails memory details = IParam.DelegationDetails(delegatee, expiry, nonce, deadline);
+        bytes memory signature = getDelegationSignature(details, router.domainSeparator(), userPrivateKey);
+        vm.expectRevert();
```
Comment (zodahu) [inline test/Router.t.sol:439]:
```suggestion
        vm.expectRevert(IRouter.InvalidNonce.selector);
```

File: test/Router.t.sol [hunk 9]
```diff
@@ -392,4 +398,178 @@ contract RouterTest is Test, LogicSignature {
...
+        vm.startPrank(user);
+        router.invalidateDelegationNonces(delegatee, newNonce);
+        (, uint128 nonce) = router.delegations(user, delegatee);
+        assertEq(nonce, newNonce);
+    }
+
+    function testCannotInvalidateExcessiveNonces() external {
+        uint128 newNonce = uint128(type(uint16).max) + 1;
+        vm.startPrank(user);
+        vm.expectRevert();
```
Comment (zodahu) [inline test/Router.t.sol:465]:
```suggestion
        vm.expectRevert(IRouter.ExcessiveInvalidation.selector);
```

File: test/Router.t.sol [hunk 10]
```diff
@@ -392,4 +398,178 @@ contract RouterTest is Test, LogicSignature {
...
+        logics[0] = IParam.Logic(
+            address(mockTo), // to
+            '',
+            inputsEmpty,
+            IParam.WrapMode.NONE,
+            address(0), // approveTo
+            address(0) // callback
+        );
+        vm.prank(delegatee);
+        vm.expectRevert();
```
Comment (zodahu) [inline test/Router.t.sol:501]:
```suggestion
        vm.expectRevert(IRouter.InvalidDelegatee.selector);
```

File: test/Router.t.sol [hunk 11]
```diff
@@ -392,4 +398,178 @@ contract RouterTest is Test, LogicSignature {
...
+            address(0) // callback
+        );
+        IParam.Fee[] memory fees = new IParam.Fee[](1);
+        fees[0] = IParam.Fee(address(mockERC20), 0, bytes32(0));
+        uint256 deadline = block.timestamp;
+        IParam.LogicBatch memory logicBatch = IParam.LogicBatch(logics, fees, deadline);
+        bytes memory signature = getLogicBatchSignature(logicBatch, router.domainSeparator(), signerPrivateKey);
+
+        vm.prank(delegatee);
+        vm.expectRevert();
```
Comment (zodahu) [inline test/Router.t.sol:571]:
```suggestion
        vm.expectRevert(IRouter.InvalidDelegatee.selector);
```

File: test/Router.t.sol [hunk 12]
```diff
@@ -392,4 +398,178 @@ contract RouterTest is Test, LogicSignature {
...
+        uint128 newNonce = 10;
+        vm.expectEmit(true, true, true, true, address(router));
+        emit NonceInvalidation(user, delegatee, newNonce, 0);
+        vm.startPrank(user);
+        router.invalidateDelegationNonces(delegatee, newNonce);
+        (, uint128 nonce) = router.delegations(user, delegatee);
+        assertEq(nonce, newNonce);
+    }
+
+    function testCannotInvalidateExcessiveNonces() external {
```
Comment (zodahu) [inline test/Router.t.sol:462]: `if (newNonce <= oldNonce) revert InvalidNonce()` hasn't been tested.

File: test/Router.t.sol [hunk 13]
```diff
@@ -392,4 +398,178 @@ contract RouterTest is Test, LogicSignature {
...
+        uint128 expiry = uint128(block.timestamp) + 3600;
+        vm.expectEmit(true, true, true, true, address(router));
+        emit Delegated(user, delegatee, expiry);
+        vm.prank(user);
+        router.allow(delegatee, expiry);
+        // (uint128 resultExpiry, uint128 resultNonce) = router.delegations(user, delegatee);
+        // assertEq(resultExpiry, expiry);
+        // assertEq(resultNonce, 0);
+        // vm.prank(delegatee);
+        // assertTrue(router.isValidDelegateeFor(user));
```
[the remark this answers is not in the document]
Comment (hihiben) [reply test/Router.t.sol:412]: Ah yes, forgot to uncomment them.

File: test/Router.t.sol (hunk 8 above)
Comment (zodahu) [reply test/Router.t.sol:439]: it's missed.

File: test/Router.t.sol [hunk 14]
```diff
@@ -392,4 +398,167 @@ contract RouterTest is Test, LogicSignature {
...
+        router.allowBySig(details, user, signature);
+        (uint128 resultExpiry, uint128 resultNonce) = router.delegations(user, delegatee);
+        assertEq(resultExpiry, expiry);
+        assertEq(resultNonce, nonce + 1);
+        vm.prank(delegatee);
+        assertTrue(router.isValidDelegateeFor(user));
+    }
+
+    function testCannotAllowBySigWithIncorrectNonce() external {
+        // Ensure correct EIP-712 encodeData
```
Comment (zodahu) [inline test/Router.t.sol:433]:
```suggestion
```

**Comment by zodahu** [comment]
LGTM

**Comment by Jeff-CCH** [comment]
LGTM

**Pull request merged**